flags.go

```diff
@@ -127,26 +161,110 @@ func Enabled(ctx context.Context, name string) (enabled bool) {
 
 	flagsMtx.RLock()
 	defer flagsMtx.RUnlock()
+	enabled = false
 	if flags == nil {
-		enabled = false
+		err = errors.New("empty flags map")
 		return
 	}
 	flag := flags[name]
-
-	// equality should be strict
-	// because Float64() can return 0
-	if f := rand.Float64(); f < flag.Rate {
-		enabled = true
+	if !flag.Active {
 		return
 	}
-	enabled = false
+	for _, r := range flag.Rules {
+
+		res := r.Handle(ctx, properties)
+		var matchBehavior string
+		if res {
+			matchBehavior = r.onMatch()
+		} else {
+			matchBehavior = r.onMiss()
+		}
+		switch matchBehavior {
+		case "on":
+			enabled = true
+			return
+		case "off":
+			enabled = false
+			return
+		case "match_list":
+			continue
+		default:
+			err = errors.New("unknown match behavior: " + matchBehavior)
+			return
+		}
+	}
+	enabled = true
 	return
 }
 
+func getProperty(ctx context.Context, props map[string]string, prop string) string {
+	if v, ok := props[prop]; ok {
+		return v
+	} else if v, ok := ctx.Value(overrideContextKey).(string); ok {
+		return v
+	} else {
+		return ""
+	}
+}
+
+func (r RateRule) Handle(ctx context.Context, props map[string]string) bool {
+	if r.Properties != nil {
+		// get the sha1 of the properties values concat
+		h := sha1.New()
+		// sort the properties for consistent behavior (should we sort on Refresh()?)
+		sort.Strings(r.Properties)
+		var buffer bytes.Buffer
+		for _, val := range r.Properties {
+			buffer.WriteString(getProperty(ctx, props, val))
+		}
+		h.Write([]byte(buffer.String()))
+		bs := h.Sum(nil)
+		encodedStr := hex.EncodeToString(bs)
```

Comment:
You can just use `binary.LittleEndian.Uint16(buf)` to get a uint16 from raw binary bytes :)

Also, we should probably use 64 bits!

Reply:
I figured there was a way instead of going thru this hex workaround! Thanks 👍 I also thought it was odd the Ruby impl limited it to 16 bits...I figured the more the better in terms of randomness/bits of entropy but tried to mirror the Ruby impl as much as possible.